### PR TITLE
[SPARK-27217][SQL] Nested schema pruning with Aggregation

### PR DESCRIPTION
**Pull request by amanomer**
### What changes were proposed in this pull request?
Added a new rule `NestColumnAliasing.Overaggregate` which will help pushdown nested columns wrapped inside `Aggregate`.

### Why are the changes needed?
Since, spark is supporting nested schema pushdown when used with `Project` (SELECT query), we also need to support same pushdown ability when user perform aggregation (such as sum) on nested columns.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Added test cases.

### CODE REVIEW

**Comment by amanomer** [comment]
cc @cloud-fan

**Comment by amanomer** [comment]
cc @cloud-fan @HyukjinKwon

**Comment by amanomer** [comment]
cc @cloud-fan @srowen @Ngone51

**Comment by amanomer** [comment]
cc @cloud-fan @HyukjinKwon

**Comment by amanomer** [comment]
@cloud-fan @HyukjinKwon kindly review this PR.

**Comment by maropu** [comment]
also, cc: @viirya @dbtsai

**Comment by maropu** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 1]
```diff
@@ -166,4 +166,48 @@ object NestedColumnAliasing {
     case _: Inline => true
     case _ => false
   }
+
+  object OverAggregate {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: `AggregateNestedColumnAliasing`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -586,8 +586,9 @@ object ColumnPruning extends Rule[LogicalPlan] {
       d.copy(child = prunedChild(child, d.references))
 
     // Prunes the unused columns from child of Aggregate/Expand/Generate/ScriptTransformation
-    case a @ Aggregate(_, _, child) if !child.outputSet.subsetOf(a.references) =>
-      a.copy(child = prunedChild(child, a.references))
+//    case a @ Aggregate(_, _, child) if !child.outputSet.subsetOf(a.references) =>
+//      a.copy(child = prunedChild(child, a.references))
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: ?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala [hunk 3]
```diff
@@ -299,6 +299,11 @@ abstract class SchemaPruningSuite
     checkAnswer(query, Row("Y.", 1) :: Row("X.", 1) :: Row(null, 2) :: Row(null, 2) :: Nil)
   }
 
+  testSchemaPruning("Spark-27217: Push nested column when used in Aggregate") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala]: super nit: plz capitalize `Spark` in the head.

**Comment by amanomer** [comment]
@cloud-fan kindly give feedback on current approach

**Comment by amanomer** [comment]
cc @cloud-fan @wangyum 
Kindly review this PR

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 4]
```diff
@@ -167,3 +169,47 @@ object NestedColumnAliasing {
...
+          }
+
+          if (nestedFieldToAlias.nonEmpty &&
+            nestedFieldToAlias.length < totalFieldNum(attr.dataType)) {
+            Some(nestedFieldToAlias)
+          } else {
+            None
+          }
+      }
+      val newProjectList: Seq[NamedExpression] =
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: This code seems be copied from `NestedColumnAliasing`. I think we can reuse the methods like `getAliasSubMap`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -586,8 +586,7 @@ object ColumnPruning extends Rule[LogicalPlan] {
       d.copy(child = prunedChild(child, d.references))
 
     // Prunes the unused columns from child of Aggregate/Expand/Generate/ScriptTransformation
-    case a @ Aggregate(_, _, child) if !child.outputSet.subsetOf(a.references) =>
-      a.copy(child = prunedChild(child, a.references))
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Why remove that? This is for top-level column pruning.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala [hunk 6]
```diff
@@ -299,6 +299,11 @@ abstract class SchemaPruningSuite
     checkAnswer(query, Row("Y.", 1) :: Row("X.", 1) :: Row(null, 2) :: Row(null, 2) :: Nil)
   }
 
+  testSchemaPruning("SPARK-27217: Push nested column when used in Aggregate") {
+    val query = sql("select sum(employer.id) from contacts")
+    checkScan(query, "struct<employer:struct<id:INT>>")
+  }
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala]: I think this is not a bug. We may not need a JIRA ticket prefix.

**Comment by viirya** [comment]
Actually I'm thinking to add nested column pruning rule for these logical operators. I think it should be feasible to have a more general one instead of adding one by one for each operator.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
@amanomer . Could you fix all UT failures?

**Comment by amanomer** [comment]
> Actually I'm thinking to add nested column pruning rule for these logical operators. I think it should be feasible to have a more general one instead of adding one by one for each operator.

I will make this rule general for other operators after resolving issues for Aggregate.

**Comment by maropu** [comment]
Yea, as @viirya said above, I also like more general one.

**Comment by dongjoon-hyun** [comment]
Gentle ping, @amanomer .

**Comment by nandini57** [comment]
Hi Spark Team, I am inclined to  add this change as a custom logical rule by copying over the Aggregate Nesting object in my spark project.We are using  Spark 2.3 version and no immediate plans to move over to 3.x. Do you think it is a viable approach ?Any guidance much appreciated

**Comment by maropu** [comment]
Probably, you'd be better to ask that in the spark mailing list. Anyway, we already have `SparkSessionExtensions` (or `SparkSession.experimental.extraOptimizations`) for injecting custom rules in 3rd-party projects. So, you can do so by using these interfaces (they are experimental interfaces though).

**Comment by nandini57** [comment]
Thanks takeshi for the quick reply.i utilized extra optimization to include
spark 4502 changes and it works pretty well.Very interested in including
27217 as well

**Comment by ntlanglois** [comment]
This optimization in aggregates would greatly benefit some of our most expensive queries against our nested schema.  We've seen up to 8x performance improvement against the same schema outside of aggregates, and to see anywhere close to this for our aggregation queries would be amazing!

There are numerous other optimizations in 3.0 that we're **very** excited for, but this SPARK-27217 seems like the only thing left that would hold back some of those optimizations from realizing their full potential in aggregations.

Thanks so much for everyone's time and work on this so far.  Just patiently wondering, @amanomer, are there are any plans to re-open this pull request with the requested changes in the near future?

**Pull request closed without merging**